visitor.php

```diff
@@ -0,0 +1,242 @@
+
+        $tagVariableType = $this->getTypeNameFromType($tagVariableType);
+
+        if ($tagVariableType === null) {
+            return null;
+        }
+
+        // It's common for an args parameter to accept a query var string or array with `string|array`.
+        // Remove the accepted string type for these so we get the strongest typing we can manage.
+        $tagVariableType = str_replace(['|string', 'string|'], '', $tagVariableType);
+
+        return sprintf(
+            " * @phpstan-param %1\$s{\n *   %2\$s,\n * } $%3\$s",
+            $tagVariableType,
+            implode(",\n *   ", $elements),
+            $tagVariableName
+        );
+    }
+
+    private function getAdditionFromReturn(Return_ $tag): ?string
+    {
+        $tagDescription = $tag->getDescription();
+        $tagVariableType = $tag->getType();
+
+        // Skip if information we need is missing.
+        if (!$tagDescription || !$tagVariableType) {
+            return null;
+        }
+
+        $elements = $this->getElementsFromDescription($tagDescription, false);
+
+        if ($elements === null) {
+            return null;
+        }
+
+        $tagVariableType = $this->getTypeNameFromType($tagVariableType);
+
+        if ($tagVariableType === null) {
+            return null;
+        }
+
+        return sprintf(
+            " * @phpstan-return %1\$s{\n *   %2\$s,\n * }",
+            $tagVariableType,
+            implode(",\n *   ", $elements)
+        );
+    }
+
+    private function getTypeNameFromType(Type $tagVariableType): ?string
+    {
+        // PHPStan dosn't support typed array shapes (`int[]{...}`) so replace
+        // typed arrays such as `int[]` with `array`.
+        $tagVariableType = preg_replace('#[a-zA-Z0-9_]+\[\]#', 'array', $tagVariableType->__toString());
+
+        if ($tagVariableType === null) {
+            return null;
+        }
+
+        if (strpos($tagVariableType, 'array') === false) {
+            // Skip if we have hash notation that's not for an array (ie. for `object`).
+            return null;
+        }
+
+        if (strpos($tagVariableType, 'array|') !== false) {
+            // Move `array` to the end of union types so the appended array shape works.
+            $tagVariableType = str_replace('array|', '', $tagVariableType) . '|array';
+        }
+
+        return $tagVariableType;
+    }
+
+    /**
+     * @return ?string[]
+     */
+    private function getElementsFromDescription(Description $tagDescription, bool $optional): ?array
+    {
+        $text = $tagDescription->__toString();
+
+        // Skip if the description doesn't contain at least one correctly
+        // formatted `@type`, which indicates an array hash.
+        if (strpos($text, '    @type ') === false) {
+            return null;
+        }
+
+        // Populate `$types` with the value of each top level `@type`.
+        $types = preg_split('/\R+    @type /', $text);
+
+        if (!$types) {
+            return null;
+        }
```

Comment:
Please avoid type juggling. `if` needs a boolean.
```suggestion
        if ($types === false) {
            return null;
        }
```
https://www.php.net/manual/en/function.preg-split.php